File: regions/system/library/swift/pagination/Sources/K1s0Pagination/PageRequest.swift
public struct PageRequest: Sendable {
    public let page: UInt32
    public let perPage: UInt32

    public init(page: UInt32, perPage: UInt32) {
        self.page = page
        self.perPage = perPage
    }
}

public struct PaginationMeta: Sendable {
    public let total: UInt64
    public let page: UInt32
    public let perPage: UInt32
    public let totalPages: UInt32

    public init(total: UInt64, page: UInt32, perPage: UInt32, totalPages: UInt32) {
        self.total = total
        self.page = page
        self.perPage = perPage
        self.totalPages = totalPages
    }
}

public enum PerPageValidationError: Error, Sendable {
    case outOfRange(value: UInt32, min: UInt32, max: UInt32)
}

public let minPerPage: UInt32 = 1
public let maxPerPage: UInt32 = 100

public func validatePerPage(_ perPage: UInt32) throws -> UInt32 {
    guard perPage >= minPerPage, perPage <= maxPerPage else {
        throw PerPageValidationError.outOfRange(value: perPage, min: minPerPage, max: maxPerPage)
    }
    return perPage
}

File: regions/system/library/csharp/pagination/PageRequest.cs
namespace K1s0.System.Pagination;

public record PageRequest(uint Page, uint PerPage);

public record PaginationMeta(long Total, uint Page, uint PerPage, uint TotalPages);

public static class PerPageValidator
{
    public const uint MinPerPage = 1;
    public const uint MaxPerPage = 100;

    public static uint Validate(uint perPage)
    {
        if (perPage < MinPerPage || perPage > MaxPerPage)
        {
            throw new ArgumentOutOfRangeException(
                nameof(perPage),
                perPage,
                $"perPage must be between {MinPerPage} and {MaxPerPage}");
        }

        return perPage;
    }
}
